src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = { ...state };
+  const history = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      for (const [key, value] of Object.entries(action.extraData)) {
+        newState[key] = value;
+      }
+      history.push({ ...newState });
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+      history.push({ ...newState });
+    }
+
+    if (action.type === 'clear') {
+      Object.keys(newState).forEach((key) => {
+        delete newState[key];
+      });
+      history.push({ ...newState });
+    }
```

Comment:
The current implementation uses multiple if statements, which can be replaced with a switch statement for better readability and adherence to coding standards when dealing with a limited amount of conditions.